### PR TITLE
fix(metro-file-map): Fix Windows cross-device path resolution

### PR DESCRIPTION
**Pull request by kitten**
## Summary

Related PR for `expo/expo` fork: https://github.com/expo/expo/pull/45648
Supersedes #1696

Cross-device path normals are intended to be relative paths from the rootDir. If the rootDir is at `C:\project` then a cross-device normal path at `D:\temp` is represented as `..\..\D:\temp`. This assumption is broken in several places, like `normalToAbsolute` printing, leading to `C:\D:\` paths that are invalid.

Changelog: [Fix] Fix Windows path handling for cross-drive paths

## Test plan

- Unit tests were adjusted to demonstrate all failing cases
- **Note:** Invalid unit tests have been removed for paths that rise above the filesystem root (they're assumed to be impossible, and their output doesn't have to be deterministic/correct)

### CODE REVIEW

**Comment by robhogan** [comment]
Thanks again for looking at this!

> Invalid unit tests have been removed for paths that rise above the filesystem root (they're assumed to be impossible, and their output doesn't have to be deterministic/correct)

This is very edge-casey but the behaviour of `..` at the root *is* actually specified (at least on posix), those test cases weren't invalid or accidental.

https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap04.html#tag_04_13

> The special filename dot shall refer to the directory specified by its predecessor. The special filename dot-dot shall refer to the parent directory of its predecessor directory. **As a special case, in the root directory, dot-dot may refer to the root directory itself.**

In particular, if you create a symlink with the target `../../../../file-at-root.txt` at `/`, it will deference to `/file-at-root.txt` on posix, so it should with Metro as well.

**Comment by robhogan** [comment]
(Also, nit: device != drive. These are cross-drive paths)

**Comment by kitten** [comment]
interesting! that's an easy fix, but requires us to use `startsWithDriveLetter` (as per https://github.com/expo/expo/pull/45648) in one place, I believe, so it's a bit of a trade-off. I didn't measure performance in that case, but I think it might not be otherwise possible to differentiate between the Windows/Posix root case otherwise. Although, I could be mistaken and there could be an easier way, if we ignore this case for Windows.

File: packages/metro-file-map/src/lib/TreeFS.js [hunk 1]
```diff
@@ -1230,10 +1240,18 @@ export default class TreeFS implements MutableFileSystem {
...
       absoluteSymlinkTarget,
     );
+    // path.relative leaves cross-device targets absolute on Windows (e.g.
+    // 'D:\\file' from a 'C:\\' root). Canonicalise those via RootPathUtils so
+    // they round-trip as a '..'-chain from rootDir.
+    if (path.isAbsolute(normalSymlinkTarget)) {
+      normalSymlinkTarget = this.#pathUtils.absoluteToNormal(
+        absoluteSymlinkTarget,
+      );
+    }
```
Comment (robhogan) [inline packages/metro-file-map/src/lib/TreeFS.js]: Any reason we need two different paths here? Why not...

```suggestion
    const normalSymlinkTarget = this.#pathUtils.absoluteToNormal(absoluteSymlinkTarget);
```

File: packages/metro-file-map/src/lib/RootPathUtils.js [hunk 2]
```diff
@@ -149,6 +152,10 @@ export class RootPathUtils {
     const right = pos === 0 ? normalPath : normalPath.slice(pos);
     if (right.length === 0) {
       return left;
+    } else if (IS_WIN32 && pos > this.#rootDepth * UP_FRAGMENT_SEP_LENGTH) {
+      // On Windows, up-fragments at the root encode target on anoher drive
+      // (e.g. '..\..\D:\file'), but on POSIX '..' re-enters the root
+      return right;
     }
```
Comment (robhogan) [inline packages/metro-file-map/src/lib/RootPathUtils.js:161]: I think this is not really a fundamental difference between OSes and more a limitation due to Metro's weird (sorry) cross-drive path encoding.

It we *could* maintain "`..` goes nowhere at the root" we would, but we can't.

```suggestion
    } else if (IS_WIN32 && pos > this.#rootDepth * UP_FRAGMENT_SEP_LENGTH) {
      // On a real file system, navigating to `..` at the top level (posix `/`
      // or Windows drive) is a no-op, but we can't respect that on Windows
      // because Metro uses e.g. `..\..\D:\foo` to represent cross-drive
      // relative paths.
      return right;
    }
```

File: packages/metro-file-map/src/lib/TreeFS.js (hunk 1 above)
Comment (kitten) [reply packages/metro-file-map/src/lib/TreeFS.js]: No reason. I think I got confused while converting this from the Expo-fork's version, where we don't go through `path.resolve` or `path.relative` anymore. This looks correct to me

Comment (kitten) [reply packages/metro-file-map/src/lib/TreeFS.js]: Actually, this seems to be failing tests. I think in Expo it doesn't because we manually normalise `target[len - 1] === path.sep` with a `slice(0, -1)` for symlink targets, which `path.relative` here does for us

Comment (kitten) [reply packages/metro-file-map/src/lib/TreeFS.js]: I think realistically, an `endsWith(path.sep)` check here is simplest, if we don't want to replicate the `path.resolve` replacement here but drop `path.relative`. There's probably other options, but they may be more intrusive

**Review by robhogan — APPROVED** [review]
